agent/resources_monitor.go

```diff
@@ -0,0 +1,132 @@
+package agent
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/agent/proto"
+	"github.com/coder/coder/v2/cli/clistat"
+	"github.com/coder/quartz"
+)
+
+func (a *agent) pushResourcesMonitoring(ctx context.Context, aAPI proto.DRPCAgentClient24) error {
+	logger := a.logger.Named("resources_monitor")
+
+	clk := quartz.NewReal()
+
+	return PushResourcesMonitoringWithConfig(ctx, logger, clk, aAPI.GetResourcesMonitoringConfiguration, aAPI.PushResourcesMonitoringUsage)
+}
+
+type (
+	ResourcesMonitorConfigurationFetcher func(ctx context.Context, params *proto.GetResourcesMonitoringConfigurationRequest) (*proto.GetResourcesMonitoringConfigurationResponse, error)
+	ResourcesMonitorDatapointsPusher     func(ctx context.Context, params *proto.PushResourcesMonitoringUsageRequest) (*proto.PushResourcesMonitoringUsageResponse, error)
+)
+
+func PushResourcesMonitoringWithConfig(ctx context.Context,
+	logger slog.Logger,
+	clk quartz.Clock,
+	configFetcher ResourcesMonitorConfigurationFetcher,
+	datapointsPusher ResourcesMonitorDatapointsPusher,
+) error {
+	config, err := configFetcher(ctx, &proto.GetResourcesMonitoringConfigurationRequest{})
+	if err != nil {
+		return xerrors.Errorf("failed to get resources monitoring configuration: %w", err)
+	}
+
+	if !config.Enabled {
+		logger.Info(ctx, "resources monitoring is disabled - skipping")
+		return nil
+	}
+
+	resourcesFetcher, err := clistat.New()
+	if err != nil {
+		return xerrors.Errorf("failed to create resources fetcher: %w", err)
+	}
+
+	datapointsQueue := NewResourcesMonitorQueue(int(config.Config.NumDatapoints))
+
+	clk.TickerFunc(ctx, time.Duration(config.Config.TickInterval*int32(time.Second)), func() error {
```

Comment:
I believe you need to call `Wait()` on the resulting value.

Reply:
From my understanding of the quartz lib , `Wait()` is called when the TickerFunc is globally finished (either the associated context expires, or one of the ticks returns an error) - and block until this moment. 

I designed this function to just start the ticker and then return , using the context to control it. 

I can change it if you see some added value, but am not sure.